XNAControls/XNAScrollPanel.cs

```diff
@@ -0,0 +1,658 @@
+    /// </summary>
+    public Point ViewSize => new()
+        {
+            X = VerticalScrollBar.Visible ? VerticalScrollBar.X : Width,
+            Y = HorizontalScrollBar.Visible ? HorizontalScrollBar.Y : Height,
+        };
+    
+    /// <summary>
+    /// Location of the viewport over the <see cref="ContentPanel"/>.
+    /// </summary>
+    public Point CurrentViewPosition
+    {
+        get => new(-CurrentContentPanelPosition.X, -CurrentContentPanelPosition.Y);
+        set => CurrentContentPanelPosition = new(-value.X, -value.Y);
+    }
+    
+    protected Rectangle ViewWindowRectangle
+    {
+        get
+        {
+            var size = ViewSize;
+            var factor = GetTotalScalingRecursive();
+            
+            return new(GetWindowPoint(),
+                new(size.X * factor, size.Y * factor));
+        }
+    }
+
+    /// <summary>
+    /// The viewport area over the <see cref="ContentPanel"/>.
+    /// </summary>
+    public Rectangle CurrentViewRectangle => new(CurrentViewPosition, ViewSize);
+    
+    /// <summary>
+    /// Indicates whether the control can be scrolled 
+    /// </summary>
+    public (bool X, bool Y) CanScroll => (
+        AllowScroll.X && ViewSize.X < ContentSize.X,
+        AllowScroll.Y && ViewSize.Y < ContentSize.Y);
+
+    #endregion
+    
+    public XNAScrollPanel(WindowManager windowManager) : base(windowManager)
+    {
+        DrawMode = ControlDrawMode.UNIQUE_RENDER_TARGET;  // so controls can be clipped
+        
+        HorizontalScrollBar = new XNAHorizontalScrollBar(WindowManager);
+        VerticalScrollBar = new XNAScrollBar(WindowManager);
+        
+        ContentPanel = new XNAPanel(WindowManager)
+        {
+            DrawBorders = false,
+        };
+        CornerPanel = new XNAPanel(WindowManager)
+        {
+            DrawBorders = false,
+            Alpha = 1.0f,
+        };
+
+        NameChanged += XNAScrollPanel_NameChanged;
+        
+        ClientRectangleUpdated += XNAScrollPanel_ClientRectangleUpdated;
+    }
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        
+        ContentPanel.ChildAdded += ContentPanel_ChildAdded;
+        ContentPanel.ChildRemoved += ContentPanel_ChildRemoved;
+        
+        HorizontalScrollBar.Scrolled += HorizontalScrollBar_Scrolled;
+        HorizontalScrollBar.MouseScrolledHorizontally += HorizontalScrollBar_MouseScrolledHorizontally;
+        // additional handler for users without horizontal scroll
+        HorizontalScrollBar.MouseScrolled += HorizontalScrollBar_MouseScrolled;
+        
+        VerticalScrollBar.Scrolled += VerticalScrollBar_Scrolled;
+        VerticalScrollBar.MouseScrolled += VerticalScrollBar_MouseScrolled;
+        
+        // black corner texture by default (accepting better solutions)
+        CornerPanel.BackgroundTexture = AssetLoader.CreateTexture(Color.Black, 2, 2);
+        CornerPanel.PanelBackgroundDrawMode = PanelBackgroundImageDrawMode.STRETCHED;
```

Comment:
Maybe optionally load a texture? Could check if `AssetLoader.AssetExists("cornerpanel.png")` and if yes, load the texture, otherwise create this black corner texture.

Reply:
Actually nevermind that, I said myself that the black texture is not an ideal solution lol